File: golang/blockchain/blockchain.go
/*
GetBlockchain : blockchain 구조체를 리턴한다. (만일 없으면 초기화한다.) + 싱글톤 패턴을 사용한다.
blockchain은 메모리 효율을 위해 블럭의 주소를 저장한다.
AddBlock : blockchain에 블럭을 추가한다. (createBlock, getLastHash, calculateHash 함수를 사용한다.)
AllBlocks : blockchain의 모든 블럭 리스트를 리턴한다.
*/

package blockchain

import (
	"crypto/sha256"
	"fmt"
	"sync"
)

type Block struct {
	Data     string
	Hash     string
	PrevHash string
}

type blockchain struct {
	blocks []*Block
}

var bc *blockchain

var once sync.Once

func (b *Block) calculateHash() {
	hash := sha256.Sum256([]byte(b.Data + b.PrevHash))
	b.Hash = fmt.Sprintf("%x", hash)
}

func (bc *blockchain) getLastHash() string {
	totalBlocks := len(bc.blocks)
	if totalBlocks == 0 {
		return ""
	}
	return bc.blocks[totalBlocks-1].Hash
}

func (bc *blockchain) createBlock(data string) *Block {
	newBlock := Block{data, "", bc.getLastHash()}
	newBlock.calculateHash()
	return &newBlock
}

func (bc *blockchain) AddBlock(data string) {
	bc.blocks = append(bc.blocks, bc.createBlock(data))
}

func GetBlockchain() *blockchain {
	if bc == nil {
		once.Do(func() {
			bc = &blockchain{}
			bc.AddBlock("Genesis Block")
		})

	}
	return bc
}

func (bc *blockchain) AllBlocks() []*Block {
	return GetBlockchain().blocks
}

File: history/1_basic_blockchain/blockchain.py
import hashlib
from typing import List, Optional


class Block:
    def __init__(self, data: str, prev_hash: str, height: int):
        self.data = data
        self.prev_hash = prev_hash
        self.height = height
        self.hash = self.calculate_hash()

    def calculate_hash(self) -> str:
        combined = self.data + self.prev_hash
        return hashlib.sha256(combined.encode("utf-8")).hexdigest()


class Blockchain:

    def __init__(self):
        self.blocks: List[Block] = []
        self.add_block("Genesis Block")

    def get_last_hash(self) -> str:
        return self.blocks[-1].hash if self.blocks else ""

    def create_block(self, data: str) -> Block:
        prev_hash = self.get_last_hash()
        height = len(self.blocks) + 1
        return Block(data, prev_hash, height)

    def add_block(self, data: str):
        block = self.create_block(data)
        self.blocks.append(block)

    def all_blocks(self) -> List[Block]:
        return self.blocks

    def get_block(self, height: int) -> Optional[Block]:
        if 0 < height <= len(self.blocks):
            return self.blocks[height - 1]
        return None


blockchain = Blockchain()
